### PR TITLE
Rename-DbaDatabase, first round

### PR DESCRIPTION
**Pull request by niphlod**
## Type of Change
 - [ ] Bug fix (non-breaking change, fixes #<enter issue number>)
 - [x] New feature (non-breaking change, adds functionality, see #289)
 - [ ] Breaking change (effects multiple commands or functionality)
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1)
 - [ ] Pester test is included

### Purpose
#289 holds pretty much everything. Not sure if this needs to handle also "moves" other than renames, such as "move this datafile to another disk".

This is just a first draft, but it's a working one nonetheless.
Cringe points:
- the templating system (I'm not good at naming things)
- general code style and loops, I feel there's a better way but can't find it
- it tries to proceed a database at a time and stop if anything goes wrong as soon as possible. Unfortunately reverting to the previous state can surely be done but could incur in the same errors, so for the moment...let's hope to not break everything
- the "default recursive behaviour" (see latest examples)
- what to return to the user, as it's a process made of several steps and ATM I can't figure a proper "layout" for the resulting object

For testers: do a backup of everything (or even better, a system image). This can (and probably will) nuclearize everything (again, "new born code alert").

I'm open to reviews and/or comments (don't go too technical on me, let's do that on round 2 or 3) ^_^

### CODE REVIEW

**Comment by potatoqualitee** [comment]
Can you guys test this one? I'll be super busy this week and only able to peep in momentarily

**Comment by gbargsley** [comment]
First batch of testing worked successfully.  My third test was to rename files.  There were some issues.  I am going to run again and see what the issue is.

image

**Comment by niphlod** [comment]
@gbargsley : fortunately it's very easy to reproduce. Scripting the structure of a database is really all is needed. Ping me back if you're experiencing issues (or if you have proposals, too ^_^)

**Comment by niphlod** [comment]
most of the cringe is due to SMO enums being reordered while cycled and renamed. Trying to figure out a way to avoid it...

**Comment by niphlod** [comment]
ok, I _think_ I found the way ( `@($whatever)` )

**Comment by gbargsley** [comment]
Does this still need some testing?

**Comment by niphlod** [comment]
The code has been refined and a new param has been added. Please tell if
the issue you were having has been resolved ;-)

Il 10 ago 2017 6:03 AM, "gbargsley" <notifications@github.com> ha scritto:

**Comment by niphlod** [comment]
this still needs to adopt shouldprocess, will do in a later release. anyone has thoughts on the general "flow" of the operations as they are right now ?

**Comment by potatoqualitee** [comment]
hey niph,
Wow, this is a beast! 🏅  I just ran it and here's what came to mind

1. The Status is not immediately understandable and reports FULL no matter what - perhaps we add a Notes field to explain?
2. Can you make Database output $db.name instead of the full SMO object or alternatively, return the full SMO object with some properties added then filter that through Select-DefaultView
3. Can we make it rename the underlying files by default - just replace dbname with newdbname and if they dont follow that rule then they can use the advanced template
4. Maybe there should be a return under terminal errors like this?

image

**Comment by niphlod** [comment]
yeah, the returned object structure is still something I'm unsure of. I added a minimal one only for cases where the underlying file rename is invoked, because it *could* waste your database. There are a lot of "return paths" which are ATM uncovered, as I am still thinking if the "general algorithm" is the way to go.
As for renaming the underlying files by default, it's a pretty bald move. It needs to drop connections, take the db offline, alter the structure, rename the files and bring it back online. As for other cmdlets whose impact is high the actual rename needing a specific param is kinda of a safeguard against accidental newbies usage.

**Comment by potatoqualitee** [comment]
We can also throw in a prompt with HighImpact to inform them of the change. I would expect rename to rename the underlying files - i was always disappointed that Microsoft didn't do that.

File: functions/Rename-DbaDatabase.ps1 [hunk 1]
```diff
@@ -0,0 +1,713 @@
...
+
+	Renames the HR database to HR2
+
+	.EXAMPLE
+	Rename-DbaDatabase -SqlInstance sqlserver2014a -Database HR -DatabaseName "dbatools_<DBN>"
+
+	Renames the HR database to dbatools_HR
+	
+	.EXAMPLE
+	Rename-DbaDatabase -SqlInstance sqlserver2014a -Database HR -DatabaseName "dbatools_<DBN>_<DATE>"
```
Comment (wsmelton) [inline functions/Rename-DbaDatabase.ps1]: Just a discussion point for ease of use...have not had a chance to pull it down to test out code itself...

I can understand the use of the `-DatabaseName` being what the database will be renamed to but any option to making that more understood? Say calling it `-Rename` or `-NewName`....I'd prefer the latter as it just makes it more self-explanatory. We use this in other commands like `Set-DbaAgentJob` and `NewLogin` for `Rename-DbaLogin`.

Comment (wsmelton) [reply functions/Rename-DbaDatabase.ps1]: Well seeing the other parameters now though I guess `DatabaseName` is the best option.

**Comment by niphlod** [comment]
Thanks for the review!
I'm actively refactoring the code almost daily to find the best way to
handle the sum of intricacies a rename requires.

Special attention of course is due to filename renames, and handling the
machinery to translate correctly sqlinstance to the actual server name in
case of move, plus testing for access of paths and so on.
I'm trying to take the most careful path for each piece of code on that
part because it could disrupt a database if something goes wrong.
 It seems that the parameter are enough to do all kind (except crazy,
maybe) operations. For the moment testing is not needed, but I welcome any
suggestion in all departments, especially:
- parameter naming
- templating "format"
- template placeholders
- return object structure



Sorry it's taking so much time

Il 02 set 2017 4:06 AM, "Shawn Melton" <notifications@github.com> ha
scritto:

*@wsmelton* commented on this pull request.
------------------------------

In functions/Rename-DbaDatabase.ps1
<https://github.com/sqlcollaborative/dbatools/pull/2030#discussion_r136686952>
:

> +	
+	.LINK
+		https://dbatools.io/Rename-DbaDatabase
+	
+	.EXAMPLE
+	Rename-DbaDatabase -SqlInstance sqlserver2014a -Database HR -DatabaseName HR2
+
+	Renames the HR database to HR2
+
+	.EXAMPLE
+	Rename-DbaDatabase -SqlInstance sqlserver2014a -Database HR
-DatabaseName "dbatools_<DBN>"
+
+	Renames the HR database to dbatools_HR
+	
+	.EXAMPLE
+	Rename-DbaDatabase -SqlInstance sqlserver2014a -Database HR
-DatabaseName "dbatools_<DBN>_<DATE>"

Well seeing the other parameters now though I guess DatabaseName is the
best option.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/sqlcollaborative/dbatools/pull/2030#discussion_r136686952>,
or mute the thread

**Comment by potatoqualitee** [comment]
Thanks so much to both of you! 💟

**Comment by niphlod** [comment]
ok, the second round is in.
Please read the docs, try it and come back with your thoughts.

Test-PSRemoting changed behaviour .... with that `-Continue` switch you don't get a $false back and everything falls apart.

**Comment by niphlod** [comment]
yeah, totally.

**Comment by Zhunya** [comment]
Got errors for Test-PSRemoting -ComputerName <servername>
```powershell
> Test-PSRemoting -Computername sedna
Write-Message : Cannot process argument transformation on parameter 'Silent'. Cannot convert value "" to type "System.Boolean". Boolean parameters accept only Boolean values and numbers, su
ch as $True, $False, 1 or 0.
At E:\GitHub\dbatools\internal\Test-PSRemoting.ps1:18 char:5
+     Write-Message -Level Verbose -Message "Starting $($MyInvocation.M ...
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Write-Message], ParameterBindingArgumentTransformationException
    + FullyQualifiedErrorId : ParameterArgumentTransformationError,Write-Message

Write-Message : Cannot process argument transformation on parameter 'Silent'. Cannot convert value "" to type "System.Boolean". Boolean parameters accept only Boolean values and numbers, su
ch as $True, $False, 1 or 0.
At E:\GitHub\dbatools\internal\Test-PSRemoting.ps1:22 char:5
+     Write-Message -Level Verbose -Message "Testing $computername"
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Write-Message], ParameterBindingArgumentTransformationException
    + FullyQualifiedErrorId : ParameterArgumentTransformationError,Write-Message

Stop-Function : Cannot process argument transformation on parameter 'Silent'. Cannot convert value "" to type "System.Boolean". Boolean parameters accept only Boolean values and numbers, su
ch as $True, $False, 1 or 0.
At E:\GitHub\dbatools\internal\Test-PSRemoting.ps1:28 char:7
+       Stop-Function -Message "Remote testing failed for computer $Com ...
+       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Stop-Function], ParameterBindingArgumentTransformationException
    + FullyQualifiedErrorId : ParameterArgumentTransformationError,Stop-Function

Write-Message : Cannot process argument transformation on parameter 'Silent'. Cannot convert value "" to type "System.Boolean". Boolean parameters accept only Boolean values and numbers, su
ch as $True, $False, 1 or 0.
At E:\GitHub\dbatools\internal\Test-PSRemoting.ps1:35 char:5
+     Write-Message -Level Verbose -Message "Ending $($MyInvocation.Myc ...
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Write-Message], ParameterBindingArgumentTransformationException
    + FullyQualifiedErrorId : ParameterArgumentTransformationError,Write-Message
```

**Comment by niphlod** [comment]
Test-psremoting "suffered" a rogue update but I'm unsure about its
correctness myself

Il 18 set 2017 5:22 PM, "Oleg Strutinskii" <notifications@github.com> ha
scritto:

**Comment by wsmelton** [comment]
@niphlod the Test-PSRemoting command needs to have the `[switch]$Silent` added to the parameter block.

**Comment by niphlod** [comment]
doesn't it need also removal of the -Continue switch on https://github.com/sqlcollaborative/dbatools/pull/2248/files#diff-ae6c90ca48f964eab94b1e64a89fb540L28 ?

**Comment by wsmelton** [comment]
You can because it is not really needed.

**Comment by niphlod** [comment]
allright, fixed the merge conflict. back to implementing shouldprocess.

**Comment by niphlod** [comment]
yeah, now I can spend the next 8 weeks to figure out tests ^_^ . to all testers .... bring it on!

**Comment by potatoqualitee** [comment]
I will test this weekend prolly 🍾

**Comment by potatoqualitee** [comment]
Hello,
Great command! I currently have issue with the output.

`Database = $db` should be `Database = $db.name` to remove the brackets ($db is good when writing verbose output)

Also, as a casual user, I do not understand what DBN, FGN, LGN, etc are. Can you make them human readable like DatabaseNameBefore, etc.

Also, can you make them show only if they are specified? `$defaultcolumns = 'Blah','Blah2','blah3'` then `if ($dbn) { $defaultcolumns += DatabaseNameBefore }` etc.

Thank you!

**Comment by potatoqualitee** [comment]
Also, would it be possible to have a -Force that renames everything - logical, physical, filegroups, etc?

**Comment by niphlod** [comment]
hugh. let's go in order then....
 - `Database = $db should be Database = $db.name`  no problem
 - DBN, FGN, LGN, FNN are for "powershell consumption". DatabaseRenames, FileGroupsRenames, LogicalNameRenames and FileNameRenames are for "human consumption". 
 - `can you make them show only if they are specified` : are you suggesting to change the output format based on the passed parameters ?
- `would it be possible to have a -Force that renames everything` yeah but based on what template ?

**Comment by potatoqualitee** [comment]
* DBN, FGN, LGN, FNN are for "powershell consumption". DatabaseRenames, FileGroupsRenames, LogicalNameRenames and FileNameRenames are for "human consumption".

Ah ha! My output ony showed DBN, FGN, LGN, FNN , not the human readable column names ones

* are you suggesting to change the output format based on the passed parameters ?

Yes, but using Select-DefaultView so that they are always there just sometimes hidden

* would it be possible to have a -Force that renames everything yeah but based on what template 

replacing the dbname and if it doesnt exist in the file then prepend it?

**Comment by niphlod** [comment]
heh! Not so fast. 
- `My output ony showed DBN, FGN, LGN, FNN` you didn't read the docs :-P 
- `changing the output based on the params` it'll make tests harder when you test for not hidden props. Also, I'm not entirely on board with a command that doesn't output the same things always, but, alas, you're the BDFL
- `replacing the dbname and if it doesnt exist in the file then prepend it` given there's a Force already for the usual things a Force is there for (killing connections) I'd go for another param, the name of which escapes me. Kinda "use this template for everything", which means basically obtaining the same result as you were passing
`.... -FileGroupName '<DBN>_<FGN>' -LogicalName '<DBN>_<LGN>' -FileName '<DBN>_<FNN>' -ReplaceBefore`

**Comment by potatoqualitee** [comment]
* My output ony showed DBN, FGN, LGN, FNN you didn't read the docs :-P

The (default) output should be obvious without reading the docs, no? I say yes ;)

* changing the output based on the params it'll make tests harder when you test for not hidden props. Also, I'm not entirely on board with a command that doesn't output the same things always, but, alas, you're the BDFL

I'm not entirely on board either. Well, at least for this command. I just know I don't like the current output and I'm not sure how to handle it. It's a whole lot of empty by default which is unusual.

* Kinda "use this template for everything", which means basically obtaining the same result as you were passing 

-Force was never explicitly set aside for disconnects but I see your point. How about `-RenameAll`? Can you provide that Template so the user doesn't have to? Personally, I don't care about my filenames, so long as they don't conflict with other dbs and make mild sense.

**Comment by niphlod** [comment]
trying to figure output props and param names is what made this command wait so much in the "pipeline forge"....... the function doesn't rename just files..... it takes care of templating renames for the db itself, the filegroup, the logicalname AND the filenames. 

Also, the "prevention mode" you're seeing in the function (i.e. not as easy-peasy, the stress in the docs about testing first, the -Preview param) is the fact that renaming filenames CAN (and will) render the db unusable if physical access to the server is not possible in any of the ways this function tries to access it... which is the case for status PARTIAL and PendingRenames not empty.

I get that SSMS botched the "rename operation" since its beginnings for most of us DBAs (that, and the fact dbatools has nothing about renames is what made this function possible), but the incredible difficulty to determine with total confidence access to the physical server running that particular db makes remote operation in renaming filenames extremely painful and - IMHO - not something the function _should_ encourage as a first run.

**Comment by niphlod** [comment]
@potatoqualitee  : can we solve "the problem" making this an underlying function for pro users and create a Rename-DbaDatabaseFile for the amusement of having something easy that just deals with renaming files and uses this as the real logic ?

**Comment by niphlod** [comment]
@potatoqualitee take a look at the last commit

**Comment by potatoqualitee** [comment]
reading

**Comment by potatoqualitee** [comment]
Yes, that sounds fab - so just db name and files right? i like that

**Comment by niphlod** [comment]
nope haha. see L822 and following. 
By default "powershell-consumable" are now hidden.
Than, any *Name parameter NOT passed to the function triggers masking its "human-consumable" output prop

**Comment by niphlod** [comment]
ggrr . gonna rebase it

**Pull request closed without merging**